packages/winpanda/extra/src/winpanda/extm/extm.py

```diff
@@ -87,13 +98,58 @@ def _exec_external_commands(self, *cmd_cl_defs):
 
         :param cmd_cl_defs: tuple(str), set of command line definitions
         """
-        msg_src = self.__class__.__name__
-
-        for cmd_cl_def in cmd_cl_defs:
+        for cmd_id, cmd_cl_def in enumerate(cmd_cl_defs, start=1):
             try:
-                cm_utl.run_external_command(cmd_cl_def)
-                LOG.debug(f'{msg_src}: External command: {cmd_cl_def}: OK')
+                cmd_run_result = cm_utl.run_external_command(cmd_cl_def)
+                LOG.debug(f'{self.msg_src}: External command: {cmd_cl_def}:'
+                          f' OK')
+                self._save_extcmd_output(str(cmd_id),
```

Comment:
The subprocess `communicate()` method needs to be called to fill in stdout and stderr correctly. Since the contents are written to files, it's better to redirect stdout and stderr directly to the files.  This will avoid buffering problems if a command happens to write more than the buffer allows.

Reply:
We do not expect huge output from helper scripts and I suppose a buffer should be auto-flushed, if writing the next chunk of data to it whould exceed its capacity, so we should be safe with the current implementation. But probably I'm missing something. Which exact buffers do you mean?

Could you please give me a hint on steps how I could reproduce the issue you are talking?

Reply:
The problem is the pipe that you set for stdout and stderr in the `run_external_command` command. It has limited capacity, so you need to read both pipes while the subprocess runs, to avoid filling the buffer. I expect to see two problems:

1. If you can start a command that outputs slowly, the current code may not capture all of it.
2. If you run a command that outputs a large file, say `Get-Content "some-large-file"`, the current code may hang, but will almost certainly not save all the data to the log file.

The simplest way to avoid this would be to always redirect the stdout and stderr to files. Since this is what you need anyway, you just need a version of `run_external_command` that accepts filenames.

Reply:
I've looked at this again, and because you use the `run` call, the pipe handling is done for you. So, this part is OK. I tested the full behavior of the `run_external_command` in #6715 and apart from bad behavior if the command times-out, it is good.

However, two points to consider:
1. Since you want to treat success and error exits the same (saving the outputs), the code would be simpler if `run_external_command` did not raise an exception when the code runs, but fails with a non-zero code. You could do this by removing the `check=True` parameter.
2. The outputs could be written directly to the `winpanda` log, to show the failure happening in-place. To understand what command is failing I need to look in the `winpanda` logs, so there appears to be little benefit to putting the command output into a separate file.

Reply:
1) JG: Since you want to treat success and error exits the same (saving the outputs), the code would be simpler if run_external_command did not raise an exception when the code runs, but fails with a non-zero code. You could do this by removing the check=True parameter.

We consider any failure of an extra package installation helper as critical, so we need to fail the entire DC/OS installation process in such a case, and this is why we need the run call to raise an exception (check=True). Thus to my mind this implementation should be OK.


2) JG: The outputs could be written directly to the winpanda log, to show the failure happening in-place. To understand what command is failing I need to look in the winpanda logs, so there appears to be little benefit to putting the command output into a separate file.

    The thing is that in debug mode (which is the only mode implemented within the Beta scope) the Winpanda application generates (in the winpanda.log file) lots of textual data, reflecting command (winpanda setup, winpanda start) execution processes along with dumps of all the package configuration/metadata artifacts (dcos-config-windows.yaml, manifests, .nssm, .extra files for every package, etc.). Amount of data for a single command run can be as big as several megabytes, and so it can be really difficult to identify/track down particular parts of the whole process or parts related to a particular issue within this big log file.
    Taking into account  everything above saving stdout/stderr from extra installation helpers to the winpanda.log appears not to be the best decision, as it would make searching throughout it even more difficult. Thus, I tend to stick to the implementation that was suggested initially.

Reply:
As I said in stand-up, I'm not planning to block this PR on these issues. I am trying to ensure that we get simple, error-free code.

My point with issue 1 is that in the calling code you need to have two points where you call `_save_extcmd_output`, one for zero return codes, and one for non-zero return codes. The Python `subprocess` code will happily return both to you in the same form (a `CompletedProcess` object) which keeps the code simpler. You can still call its `check_returncode` method after writing the logs.

With issue 2, the length of individual log files is not really a problem (not when disk usage will be the same either way). More important is that diagnosing problems is easy, and requiring the user to change files to check information adds extra work for them.